# Capstone Project - The Battle of the Neighborhoods (Week 2)
### Applied Data Science Capstone by IBM/Coursera

# The Battle of Neighbourhoods | Proposal |Tamil Nadu

#Contents:

1.<a href='#intro'>Introduction to Chennai</a><br>
2.<a href='#problem_statement'>Problem Statement</a><br>
3.<a href='#location'>Location</a> <br>
4.<a href='#api'>Four Square API</a><br>
5.<a href='#lib'>Libraries Used</a> <br>
6.<a href='#data'>Data</a> <br>
7.<a href='#method'>Methodology</a> <br>
8.<a href='#results'>Results</a> <br>
9.<a href='#discussion'>Discussion</a> <br>
10.<a href='#conclusion'>Conclusion</a> <br>

# Introduction<a id='intro'></a>


Tamil Nadu is an upcoming state in terms of work opportunities and people are coming from different parts of the country to settle down in Tamil Nadu. 

This project is for those people who are looking for better neighborhoods. For ease of accessing to Cafe, School, Super market, medical shops, grocery shops, mall, theatre, hospital, like minded people, etc.

This Project aim to create an analysis of features for a people migrating to Tamil Nadu(TN) to search a best neighborhood as a comparative analysis between neighborhoods. The features include median housing price and better school according to ratings, crime rates of that particular area, road connectivity, weather conditions, good management for emergency, water resources both freash and waste water and excrement conveyed in sewers and recreational facilities.

It will help people to get awareness of the area and neighborhood before moving to a new city, state, country or place for their work or to start a new fresh life.

# Problem Statement<a id='problem_statement'></a>

The major purpose of this project, is to suggest a better neighborhood in a new city for the person who are shiffting there. Social presence in society in terms of like minded people. Connectivity to the airport, bus stand, city center, markets and other daily needs things nearby.

Sorted list of house in terms of housing prices in a ascending or descending order
Sorted list of schools in terms of location, fees, rating and reviews

# The Location:<a id='location'></a>

(TN) is an upcoming destination where people are coming from different parts of the country to settle down. We will be analyzing TN and then Chennai to find out the good neighborhood(s) to settle down.

# Foursquare API: <a id='api'></a>

This project would use Four-square API as its prime data gathering source as it has a database of millions of places, especially their places API which provides the ability to perform location search, location sharing and details about a business.

# Libraries Used<a id='lib'></a>

Pandas: For creating and manipulating dataframes.

Folium: Python visualization library would be used to visualize the neighborhoods cluster distribution of using interactive leaflet map.

Scikit Learn: For importing k-means clustering.

JSON: Library to handle JSON files.

XML: To separate data from presentation and XML stores data in plain text format.

Geocoder: To retrieve Location Data.

Beautiful Soup and Requests: To scrap and library to handle http requests.

Matplotlib: Python Plotting Module.

In [20]:
!pip install beautifulsoup4
!pip install lxml
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation


# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 


from IPython.display import display_html
import pandas as pd
import numpy as np
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe


import folium # plotting library
from bs4 import BeautifulSoup
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

print('Folium installed')
print('Libraries imported.')

You should consider upgrading via the 'C:\Users\anand\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.


You should consider upgrading via the 'C:\Users\anand\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.


Folium installed
Libraries imported.


# Data<a id='data'></a>

### Scrapping data for both Tamil Nadu and Chennai using Beautiful Soup

In [4]:
url_tn = 'https://en.wikipedia.org/wiki/List_of_districts_of_Tamil_Nadu'
url_chennai ='https://en.wikipedia.org/wiki/List_of_neighbourhoods_of_Chennai'
table_class="wikitable sortable jquery-tablesorter"
page = requests.get(url_chennai)

soup = BeautifulSoup(page.text, 'html.parser')
soup

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of neighbourhoods of Chennai - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"10e4169b-32f5-4de3-aa1b-f31655aafd02","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_neighbourhoods_of_Chennai","wgTitle":"List of neighbourhoods of Chennai","wgCurRevisionId":1025461911,"wgRevisionId":1025461911,"wgArticleId":9736473,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Use dmy dates from February 2019","Use Indian English from February 2019","All Wikipedia articles written 

In [5]:
soup = BeautifulSoup(page.text, 'html.parser')
chennaitable=soup.find('table',{'class':"wikitable"})

In [7]:
df=pd.read_html(str(chennaitable))
# convert list to dataframe
df=pd.DataFrame(df[0])
print(df.head())

              Area                Location  Latitude  Longitude
0       Adambakkam  South and East Chennai   12.9880    80.2047
1            Adyar  South and East Chennai   13.0012    80.2565
2          Alandur  South and East Chennai   12.9975    80.2006
3        Alapakkam            West Chennai   13.0490    80.1673
4  Alwarthirunagar            West Chennai   13.0426    80.1840


In [31]:
df.rename(columns ={'Location': 'Neighborhood'}, inplace = True)

In [32]:
df.head()

Area            Neighborhood   Latitude  Longitude
0       Adambakkam  South and East Chennai  12.982221  80.209121
1            Adyar  South and East Chennai  13.006450  80.257779
2          Alandur  South and East Chennai  13.004256  80.201453
3        Alapakkam            West Chennai  13.049901  80.165435
4  Alwarthirunagar            West Chennai  13.047486  80.183669

In [12]:
page2 = requests.get(url_tn)

soup = BeautifulSoup(page2.text, 'html.parser')
soup

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of districts of Tamil Nadu - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"1936cded-2e7b-4c72-a19b-42c5abb40af0","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_districts_of_Tamil_Nadu","wgTitle":"List of districts of Tamil Nadu","wgCurRevisionId":1028003208,"wgRevisionId":1028003208,"wgArticleId":1114401,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description matches Wikidata","Use Indian English from April 2020","All 

In [13]:
soup2 = BeautifulSoup(page2.text, 'html.parser')
tn_table=soup.find('table',{'class':"wikitable"})

In [17]:
df2=pd.read_html(str(tn_table))
# convert list to dataframe
df2=pd.DataFrame(df2[0])
print(df2.head())
df2.to_csv(r'C:\Users\anand\Desktop\Coursera- Capstone Project\tndata.csv', index = False)

   No.      District Code       Capital Date of official formation  \
0  1.0      Ariyalur   AR      Ariyalur           23 November 2007   
1  2.0  Chengalpattu  CGL  Chengalpattu           29 November 2019   
2  3.0       Chennai   CH       Chennai            1 November 1956   
3  4.0    Coimbatore   CO    Coimbatore            1 November 1956   
4  5.0     Cuddalore   CU     Cuddalore          30 September 1993   

                                          Split from Area (km2)  \
0                                         Perambalur    1949.31   
1                                        Kanchipuram    2944.96   
2  One of the original 13 districts (under former...        426   
3                   One of the original 13 districts  4,723[36]   
4                               South Arcot District       3703   

   Population[34]  Populationdensity(/km2)  \
0          754894                      390   
1         2556244                      868   
2         4646732                    2

# Taking the data from a csv file

In [142]:
df = pd.read_csv (r'C:\Users\anand\Desktop\Coursera- Capstone Project\Chennai Data.csv')
df2 = pd.read_csv(r'C:\Users\anand\Desktop\Coursera- Capstone Project\Capstone Project - Areas of Tamil Nadu.csv')
df2.head()


No.          Area Code  Neighborhood Area (km2) Population[34] Population
0    1      Ariyalur   AR      Ariyalur    1949.31       7,54,894        390
1    2  Chengalpattu  CGL  Chengalpattu   2,944.96      25,56,244        868
2    3       Chennai   CH       Chennai        426      46,46,732     26,076
3    4    Coimbatore   CO    Coimbatore  4,723[36]      34,58,045        732
4    5     Cuddalore   CU     Cuddalore      3,703      26,05,914        709

# # dropping all duplicates from Chennai data frame

In [18]:
df.drop_duplicates(subset ="Area",
                     keep = False, inplace = True)
df.shape

(164, 4)

In [21]:
to_drop_unknown = []
geolocator = Nominatim(user_agent="ny_explorer")
for index, row in df.iterrows():
    address = row['Area'] + ', Chennai'
    try:
        location = geolocator.geocode(address)
        latitude = location.latitude
        longitude = location.longitude
        print('The geograpical coordinate of {} are {}, {}.'.format(address, latitude, longitude))
        df.loc[index, 'Latitude'] = latitude
        df.loc[index, 'Longitude'] = longitude
    except AttributeError:
        print('Cannot do: {}, will drop index: {}'.format(address, index))
        to_drop_unknown.append(index)

The geograpical coordinate of Adambakkam, Chennai are 12.9822215, 80.209121.
The geograpical coordinate of Adyar, Chennai are 13.00645, 80.2577791.
The geograpical coordinate of Alandur, Chennai are 13.0042558, 80.2014525.
The geograpical coordinate of Alapakkam, Chennai are 13.049901, 80.1654347.
The geograpical coordinate of Alwarthirunagar, Chennai are 13.047486, 80.1836687.
The geograpical coordinate of Ambattur, Chennai are 13.1193746, 80.1507648.
The geograpical coordinate of Aminjikarai, Chennai are 13.0721399, 80.2205453.
The geograpical coordinate of Anna Nagar, Chennai are 13.0872004, 80.2164421.
The geograpical coordinate of Annanur, Chennai are 13.1163563, 80.1269127.
The geograpical coordinate of Arumbakkam, Chennai are 13.074371, 80.2081312.
The geograpical coordinate of Ashok Nagar, Chennai are 13.0400731, 80.2159247.
The geograpical coordinate of Avadi, Chennai are 13.083553, 80.233896.
The geograpical coordinate of Ayappakkam, Chennai are 12.5057641, 80.1129406.
The ge

The geograpical coordinate of Saidapet, Chennai are 13.020817, 80.2239536.
The geograpical coordinate of Saligramam, Chennai are 13.0583406, 80.1985327.
The geograpical coordinate of Selaiyur, Chennai are 12.9297073, 80.1186637.
Cannot do: Selavoyal, Chennai, will drop index: 117
The geograpical coordinate of Sembiam, Chennai are 13.1128208, 80.2388201.
The geograpical coordinate of Sholinganallur, Chennai are 12.8935356, 80.2320634.
The geograpical coordinate of Sowcarpet, Chennai are 13.0884967, 80.2783092.
Cannot do: Surapet, Chennai, will drop index: 121
The geograpical coordinate of T. Nagar, Chennai are 13.0294483, 80.2309064.
The geograpical coordinate of T.V.K. Nagar, Chennai are 11.9293204, 79.8075568.
The geograpical coordinate of Tambaram, Chennai are 12.9383611, 80.2051631.
The geograpical coordinate of Taramani, Chennai are 12.9849067, 80.2404069.
The geograpical coordinate of Teynampet, Chennai are 13.0443237, 80.2498455.
The geograpical coordinate of Thirumangalam, Chenn

# to find geographical coordinates for each city

In [40]:

geolocator = Nominatim(user_agent="ny_explorer")
for index, row in df2.iterrows():
    address = row['District'] + ', Tamil Nadu'
    try:
        location = geolocator.geocode(address)
        latitude = location.latitude
        longitude = location.longitude
        print('The geograpical coordinate of {} are {}, {}.'.format(address, latitude, longitude))
        df2.loc[index, 'Latitude'] = latitude
        df2.loc[index, 'Longitude'] = longitude
    except AttributeError:
        print('Cannot do: {}, will drop index: {}'.format(address, index))
        to_drop_unknown.append(index)

The geograpical coordinate of Ariyalur, Tamil Nadu are 11.076035950000001, 79.11745538182738.
The geograpical coordinate of Chengalpattu, Tamil Nadu are 12.6840886, 79.9836371.
The geograpical coordinate of Chennai, Tamil Nadu are 13.0836939, 80.270186.
The geograpical coordinate of Coimbatore, Tamil Nadu are 11.0018115, 76.9628425.
The geograpical coordinate of Cuddalore, Tamil Nadu are 11.7564329, 79.7634644.
The geograpical coordinate of Dharmapuri, Tamil Nadu are 12.09680475, 78.19304301026716.
The geograpical coordinate of Dindigul, Tamil Nadu are 10.3303299, 78.0673979084697.
The geograpical coordinate of Erode, Tamil Nadu are 11.369204400000001, 77.67662686841793.
The geograpical coordinate of Kallakurichi, Tamil Nadu are 11.7406259, 78.9637131.
The geograpical coordinate of Kanchipuram, Tamil Nadu are 12.836393, 79.7053304.
The geograpical coordinate of Kanyakumari, Tamil Nadu are 8.079252, 77.5499338.
The geograpical coordinate of Karur, Tamil Nadu are 10.9301522, 78.084854545

In [41]:
df.columns = df.columns.astype(str)
df.reset_index(drop=True, inplace=True)
df2.head()

No.      District Code       Capital Date of official formation  \
0  1.0      Ariyalur   AR      Ariyalur           23 November 2007   
1  2.0  Chengalpattu  CGL  Chengalpattu           29 November 2019   
2  3.0       Chennai   CH       Chennai            1 November 1956   
3  4.0    Coimbatore   CO    Coimbatore            1 November 1956   
4  5.0     Cuddalore   CU     Cuddalore          30 September 1993   

                                          Split from Area (km2)  \
0                                         Perambalur    1949.31   
1                                        Kanchipuram    2944.96   
2  One of the original 13 districts (under former...        426   
3                   One of the original 13 districts  4,723[36]   
4                               South Arcot District       3703   

   Population[34]  Populationdensity(/km2)  \
0          754894                      390   
1         2556244                      868   
2         4646732                    26076   
3         3458045                      732   
4         2605914                      709   

                                          Taluks[35]  Map   Latitude  \
0          Andimadam Ariyalur Udayarpalayam Sendurai  NaN  11.076036   
1  Chengalpattu Cheyyur Madurantakam Pallavaram T...  NaN  12.684089   
2  Alandur Ambattur Aminjikarai Ayanavaram Egmore...  NaN  13.083694   
3  Anaimalai Annur Coimbatore-North Coimbatore-So...  NaN  11.001812   
4  Bhuvanagiri Chidambaram Cuddalore Kattumannark...  NaN  11.756433   

   Longitude  
0  79.117455  
1  79.983637  
2  80.270186  
3  76.962842  
4  79.763464

# Define Foursquare Credentials and Version
 




In [23]:

# define Foursquare Credentials and Version
CLIENT_ID = 'HJSAJXEN31RXCGMHCDVIPCJHXO5XRO4ZQJK0BAU1CYK2HO4I' # your Foursquare ID
CLIENT_SECRET = 'SNVHO1XY34QM11ZQS0DH1NLP1MRUMYQMIEVVWH2PDC3HIFHW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


Your credentails:
CLIENT_ID: HJSAJXEN31RXCGMHCDVIPCJHXO5XRO4ZQJK0BAU1CYK2HO4I
CLIENT_SECRET:SNVHO1XY34QM11ZQS0DH1NLP1MRUMYQMIEVVWH2PDC3HIFHW


# Methodology <a id= 'method'> </a>


# Analyzing all districts in Tamil Nadu

In [101]:
# finding the geographical coordinates of Tamil Nadu using nominatim
address = 'TamilNadu, TN'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Tamil Nadu  are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Tamil Nadu  are 8.805977, 78.1553668.


In [51]:
# create map of Chennai using latitude and longitude values
map_tn = folium.Map(location=[latitude, longitude], zoom_start=6)

# add markers to map
for lat, lng, district, capital in zip(df2['Latitude'], df2['Longitude'], df2['District'], df2['Capital']):
    label = '{}, {}'.format(capital, district)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tn)  
map_tn

Chennai is a famous area in Tamil Nadu and let's try and visualize and cluster the neighborhoods in Chennai.So let's use areas of chennai as available in the data frame df to find out the popular neighborhood. We will be using the coordinates to find amenities and recreation facilities nearby

In [52]:
df.head()

Area            Neighborhood   Latitude  Longitude
0       Adambakkam  South and East Chennai  11.076036  79.117455
1            Adyar  South and East Chennai  12.684089  79.983637
2          Alandur  South and East Chennai  13.083694  80.270186
3        Alapakkam            West Chennai  11.001812  76.962842
4  Alwarthirunagar            West Chennai  11.756433  79.763464

Let's get the geographical coordinates of Chennai.


In [53]:
address = 'Chennai, TN'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 13.0836939, 80.270186.


Visualizing Chennai Region

In [54]:
# create map of Manhattan using latitude and longitude values
map_chennai = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df['Latitude'], df['Longitude'], df['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_chennai)  
    
map_chennai

Let's explore the first neighborhood in our dataframe

In [70]:
df.loc[7, 'Neighborhood']

'West Chennai'

Get the neighborhood's latitude and longitude values.

In [71]:
neighborhood_latitude = df.loc[7, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[7, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[7, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of West Chennai are 11.369204400000001, 77.67662686841793.


Lets explore the neighborhood in and around Chennai

In [72]:
# type your answer here
LIMIT = 300 # limit of number of venues returned by Foursquare API

radius = 5000 # define radius


url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=HJSAJXEN31RXCGMHCDVIPCJHXO5XRO4ZQJK0BAU1CYK2HO4I&client_secret=SNVHO1XY34QM11ZQS0DH1NLP1MRUMYQMIEVVWH2PDC3HIFHW&v=20180605&ll=11.369204400000001,77.67662686841793&radius=5000&limit=300'

In [73]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '60c73836948d0065fb9ac51b'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Current map view',
  'headerFullLocation': 'Current map view',
  'headerLocationGranularity': 'unknown',
  'totalResults': 9,
  'suggestedBounds': {'ne': {'lat': 11.414204445000045,
    'lng': 77.72244190932575},
   'sw': {'lat': 11.324204354999956, 'lng': 77.6308118275101}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4f11941ee4b0253d4cd5722d',
       'name': 'Liquidzz',
       'location': {'crossStreet': 'Perundurai Road',
        'lat': 11.333921642438693,
        'lng': 77.70502966134936,
        'labeledLatLngs': [{'label': 'display',
          'lat': 11.3339216

In [74]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Getting the nearby venues

In [75]:
from pandas.io.json import json_normalize
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-75-04b8cafc50e9>:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name         categories        lat        lng
0        Liquidzz  Food & Drink Shop  11.333922  77.705030
1  Domino's Pizza        Pizza Place  11.330939  77.699881
2   Axis Bank ATM                ATM  11.372102  77.703102
3   Axis Bank ATM                ATM  11.360070  77.702510
4            Kdfc        Coffee Shop  11.367552  77.647340

In [76]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

9 venues were returned by Foursquare.


# Exploring Neighborhoods around Chennai

#### Let's create a function to repeat the same process to all the neighborhoods in Chennai


In [77]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius=5000&limit=300'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [79]:
chennai_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

South and East Chennai
South and East Chennai
South and East Chennai
West Chennai
West Chennai
West Chennai
West Chennai
West Chennai
West Chennai
West Chennai
West Chennai
West Chennai
West Chennai
North Chennai
South and East Chennai
North Chennai
North Chennai
South and East Chennai
North Chennai
West Chennai
West Chennai
South and East Chennai
North Chennai
North Chennai
West Chennai
South and East Chennai
South and East Chennai
South and East Chennai
North Chennai
South and East Chennai
South and East Chennai
West Chennai
North Chennai
West Chennai
South and East Chennai
North Chennai
West Chennai
North Chennai
North Chennai
South and East Chennai
North Chennai
South and East Chennai
North Chennai
West Chennai
North Chennai
North Chennai
South and East Chennai
South and East Chennai
West Chennai
North Chennai
West Chennai
West Chennai
North Chennai
North Chennai
North Chennai
South and East Chennai
West Chennai
South and East Chennai
North Chennai
North Chennai
West Chennai
West C

In [80]:
print(chennai_venues.shape)
chennai_venues.head()

(7048, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  South and East Chennai              12.684089               79.983637   
1  South and East Chennai              12.684089               79.983637   
2  South and East Chennai              12.684089               79.983637   
3  South and East Chennai              12.684089               79.983637   
4  South and East Chennai              12.684089               79.983637   

                        Venue  Venue Latitude  Venue Longitude  \
0               Latha Cinemas       12.680661        79.980511   
1      Changalpettu Bus Stand       12.692468        79.979310   
2   Chengalpet To Beach Train       12.693512        79.981451   
3  Chengalpat Railway Station       12.704485        79.981123   
4                Kolavai Lake       12.710869        79.980555   

       Venue Category  
0       Movie Theater  
1         Bus Station  
2  Light Rail Station  
3       Train Station  
4                Lake

    Let's check how many venues were returned for each neighborhood


In [81]:
chennai_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                  
North Chennai                                          1900   
Northern Suburbs of Chennai                              14   
South and East Chennai                                 2611   
Southern-Eastern Suburbs of Chennai                     156   
Suburban Chennai                                        143   
Suburbs along ECR and OMR of Chennai                    204   
West Chennai                                           1992   
Western Suburbs of Chennai                               28   

                                      Neighborhood Longitude  Venue  \
Neighborhood                                                          
North Chennai                                           1900   1900   
Northern Suburbs of Chennai                               14     14   
South and East Chennai                                  2611   2611   
Southern-Eastern Suburbs of Chennai                      156    156   
Suburban Chennai                                         143    143   
Suburbs along ECR and OMR of Chennai                     204    204   
West Chennai                                            1992   1992   
Western Suburbs of Chennai                                28     28   

                                      Venue Latitude  Venue Longitude  \
Neighborhood                                                            
North Chennai                                   1900             1900   
Northern Suburbs of Chennai                       14               14   
South and East Chennai                          2611             2611   
Southern-Eastern Suburbs of Chennai              156              156   
Suburban Chennai                                 143              143   
Suburbs along ECR and OMR of Chennai             204              204   
West Chennai                                    1992             1992   
Western Suburbs of Chennai                        28               28   

                                      Venue Category  
Neighborhood                                          
North Chennai                                   1900  
Northern Suburbs of Chennai                       14  
South and East Chennai                          2611  
Southern-Eastern Suburbs of Chennai              156  
Suburban Chennai                                 143  
Suburbs along ECR and OMR of Chennai             204  
West Chennai                                    1992  
Western Suburbs of Chennai                        28

#### Let's find out how many unique categories can be curated from all the returned venues


In [82]:
print('There are {} uniques categories.'.format(len(chennai_venues['Venue Category'].unique())))

There are 187 uniques categories.


# Analyzing the neighborhoods

In [83]:
# one hot encoding
chennai_onehot = pd.get_dummies(chennai_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
chennai_onehot['Neighborhood'] = chennai_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [chennai_onehot.columns[-1]] + list(chennai_onehot.columns[:-1])
chennai_onehot = chennai_onehot[fixed_columns]

chennai_onehot.head()

Neighborhood  ATM  Accessories Store  Afghan Restaurant  \
0  South and East Chennai    0                  0                  0   
1  South and East Chennai    0                  0                  0   
2  South and East Chennai    0                  0                  0   
3  South and East Chennai    0                  0                  0   
4  South and East Chennai    0                  0                  0   

   African Restaurant  Airport Lounge  Airport Terminal  American Restaurant  \
0                   0               0                 0                    0   
1                   0               0                 0                    0   
2                   0               0                 0                    0   
3                   0               0                 0                    0   
4                   0               0                 0                    0   

   Amphitheater  Andhra Restaurant  ...  Train  Train Station  \
0             0                  0  ...      0              0   
1             0                  0  ...      0              0   
2             0                  0  ...      0              0   
3             0                  0  ...      0              1   
4             0                  0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Store  Vietnamese Restaurant  \
0                              0            0                      0   
1                              0            0                      0   
2                              0            0                      0   
3                              0            0                      0   
4                              0            0                      0   

   Waterfall  Wine Shop  Women's Store  Zoo  Zoo Exhibit  
0          0          0              0    0            0  
1          0          0              0    0            0  
2          0          0              0    0            0  
3          0          0              0    0            0  
4          0          0              0    0            0  

[5 rows x 188 columns]

In [84]:
chennai_onehot.shape

(7048, 188)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [85]:
chennai_grouped = chennai_onehot.groupby('Neighborhood').mean().reset_index()
chennai_grouped

Neighborhood       ATM  Accessories Store  \
0                         North Chennai  0.001579           0.001053   
1           Northern Suburbs of Chennai  0.000000           0.000000   
2                South and East Chennai  0.001532           0.000000   
3   Southern-Eastern Suburbs of Chennai  0.000000           0.000000   
4                      Suburban Chennai  0.000000           0.000000   
5  Suburbs along ECR and OMR of Chennai  0.000000           0.000000   
6                          West Chennai  0.003514           0.000502   
7            Western Suburbs of Chennai  0.000000           0.000000   

   Afghan Restaurant  African Restaurant  Airport Lounge  Airport Terminal  \
0           0.001053            0.001579        0.000000          0.000000   
1           0.000000            0.000000        0.000000          0.000000   
2           0.000383            0.002298        0.010341          0.002298   
3           0.000000            0.006410        0.000000          0.000000   
4           0.000000            0.000000        0.000000          0.000000   
5           0.000000            0.000000        0.000000          0.000000   
6           0.000502            0.000502        0.006024          0.000502   
7           0.000000            0.000000        0.000000          0.000000   

   American Restaurant  Amphitheater  Andhra Restaurant  ...     Train  \
0             0.002632      0.000000           0.000000  ...  0.003158   
1             0.000000      0.000000           0.000000  ...  0.000000   
2             0.000766      0.000383           0.001149  ...  0.000000   
3             0.000000      0.000000           0.000000  ...  0.000000   
4             0.000000      0.000000           0.000000  ...  0.000000   
5             0.009804      0.000000           0.000000  ...  0.000000   
6             0.004518      0.000000           0.005020  ...  0.000000   
7             0.000000      0.000000           0.000000  ...  0.000000   

   Train Station  Vegetarian / Vegan Restaurant  Video Store  \
0       0.029474                       0.017895     0.008421   
1       0.357143                       0.000000     0.000000   
2       0.015703                       0.024129     0.000766   
3       0.000000                       0.019231     0.000000   
4       0.013986                       0.020979     0.006993   
5       0.000000                       0.000000     0.000000   
6       0.004518                       0.029116     0.000000   
7       0.035714                       0.035714     0.000000   

   Vietnamese Restaurant  Waterfall  Wine Shop  Women's Store       Zoo  \
0               0.001053   0.000000   0.001053       0.000526  0.000000   
1               0.000000   0.000000   0.000000       0.000000  0.000000   
2               0.000000   0.000383   0.000000       0.000000  0.000766   
3               0.000000   0.000000   0.000000       0.000000  0.012821   
4               0.000000   0.000000   0.000000       0.000000  0.000000   
5               0.000000   0.000000   0.000000       0.000000  0.004902   
6               0.000502   0.000000   0.000502       0.001004  0.000000   
7               0.000000   0.000000   0.000000       0.000000  0.000000   

   Zoo Exhibit  
0     0.000000  
1     0.000000  
2     0.000766  
3     0.012821  
4     0.000000  
5     0.000000  
6     0.000000  
7     0.000000  

[8 rows x 188 columns]

In [86]:
chennai_grouped.shape

(8, 188)

#### Let's print each neighborhood along with the top 5 most common venues


In [87]:
num_top_venues = 40

for hood in chennai_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = chennai_grouped[chennai_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----North Chennai----
                            venue  freq
0               Indian Restaurant  0.17
1                           Hotel  0.05
2                     Pizza Place  0.05
3            Fast Food Restaurant  0.05
4                       Multiplex  0.04
5                     Coffee Shop  0.04
6                            Café  0.04
7              Italian Restaurant  0.03
8                   Train Station  0.03
9                       Bookstore  0.02
10                      Juice Bar  0.02
11                  Movie Theater  0.02
12                 Ice Cream Shop  0.02
13                 Sandwich Place  0.02
14             Chinese Restaurant  0.02
15               Department Store  0.02
16  Vegetarian / Vegan Restaurant  0.02
17                            Bar  0.01
18                     Restaurant  0.01
19                 Clothing Store  0.01
20                  Shopping Mall  0.01
21                 Cricket Ground  0.01
22                         Market  0.01
23                

Let's put that into a pandas dataframe


In [88]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [89]:
num_top_venues = 15

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = chennai_grouped['Neighborhood']

for ind in np.arange(chennai_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(chennai_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                        North Chennai     Indian Restaurant   
1          Northern Suburbs of Chennai         Train Station   
2               South and East Chennai     Indian Restaurant   
3  Southern-Eastern Suburbs of Chennai     Indian Restaurant   
4                     Suburban Chennai     Indian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Fast Food Restaurant           Pizza Place                 Hotel   
1                   Bar                  Pier     Indian Sweet Shop   
2           Pizza Place  Fast Food Restaurant                  Café   
3                  Café           Pizza Place                 Hotel   
4                  Café  Fast Food Restaurant           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Café           Coffee Shop             Multiplex   
1             Surf Spot                  Farm              Tea Room   
2                 Hotel    Chinese Restaurant           Coffee Shop   
3                Bakery           Coffee Shop        Ice Cream Shop   
4           Pizza Place             Multiplex            Restaurant   

  8th Most Common Venue 9th Most Common Venue         10th Most Common Venue  \
0    Italian Restaurant         Train Station                 Sandwich Place   
1           Snack Place                 Beach                     Food Court   
2            Restaurant    Italian Restaurant                 Ice Cream Shop   
3      Department Store        Sandwich Place                      Juice Bar   
4    Italian Restaurant                 Hotel  Vegetarian / Vegan Restaurant   

  11th Most Common Venue         12th Most Common Venue  \
0         Ice Cream Shop               Department Store   
1                    ATM                      Multiplex   
2              Multiplex  Vegetarian / Vegan Restaurant   
3     Chinese Restaurant                     Restaurant   
4         Ice Cream Shop             Chinese Restaurant   

  13th Most Common Venue         14th Most Common Venue  \
0              Bookstore  Vegetarian / Vegan Restaurant   
1                 Museum                    Music Store   
2         Sandwich Place                   Burger Joint   
3   Fast Food Restaurant                 Clothing Store   
4              Bookstore                Thai Restaurant   

    15th Most Common Venue  
0                Juice Bar  
1  New American Restaurant  
2                    Beach  
3       Italian Restaurant  
4           Sandwich Place

#Performing kmeans clustering

In [90]:
# set number of clusters
kclusters = 4

chennai_grouped_clustering = chennai_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(chennai_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 2, 3, 3, 3, 1, 3, 0])

In [95]:
chennai_merged = df

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
chennai_merged = chennai_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

chennai_merged.head() # check the last columns!

Area            Neighborhood   Latitude  Longitude  \
0       Adambakkam  South and East Chennai  11.076036  79.117455   
1            Adyar  South and East Chennai  12.684089  79.983637   
2          Alandur  South and East Chennai  13.083694  80.270186   
3        Alapakkam            West Chennai  11.001812  76.962842   
4  Alwarthirunagar            West Chennai  11.756433  79.763464   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               3     Indian Restaurant           Pizza Place   
1               3     Indian Restaurant           Pizza Place   
2               3     Indian Restaurant           Pizza Place   
3               3     Indian Restaurant           Pizza Place   
4               3     Indian Restaurant           Pizza Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0  Fast Food Restaurant                  Café                 Hotel   
1  Fast Food Restaurant                  Café                 Hotel   
2  Fast Food Restaurant                  Café                 Hotel   
3  Fast Food Restaurant                  Café        Sandwich Place   
4  Fast Food Restaurant                  Café        Sandwich Place   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Chinese Restaurant           Coffee Shop            Restaurant   
1    Chinese Restaurant           Coffee Shop            Restaurant   
2    Chinese Restaurant           Coffee Shop            Restaurant   
3                 Hotel           Coffee Shop        Clothing Store   
4                 Hotel           Coffee Shop        Clothing Store   

  9th Most Common Venue         10th Most Common Venue 11th Most Common Venue  \
0    Italian Restaurant                 Ice Cream Shop              Multiplex   
1    Italian Restaurant                 Ice Cream Shop              Multiplex   
2    Italian Restaurant                 Ice Cream Shop              Multiplex   
3        Ice Cream Shop  Vegetarian / Vegan Restaurant              Multiplex   
4        Ice Cream Shop  Vegetarian / Vegan Restaurant              Multiplex   

          12th Most Common Venue 13th Most Common Venue  \
0  Vegetarian / Vegan Restaurant         Sandwich Place   
1  Vegetarian / Vegan Restaurant         Sandwich Place   
2  Vegetarian / Vegan Restaurant         Sandwich Place   
3             Chinese Restaurant             Restaurant   
4             Chinese Restaurant             Restaurant   

  14th Most Common Venue 15th Most Common Venue  
0           Burger Joint                  Beach  
1           Burger Joint                  Beach  
2           Burger Joint                  Beach  
3     Italian Restaurant       Asian Restaurant  
4     Italian Restaurant       Asian Restaurant

Plotting the clusters using Folium

In [100]:
import matplotlib.cm as cm
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=15)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(chennai_merged['Latitude'], chennai_merged['Longitude'], chennai_merged['Neighborhood'], chennai_merged['Cluster Labels']):
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#<b>finding out the best neighborhood in Chennai</b>

In [99]:
chennai_merged.loc[chennai_merged['Cluster Labels'] == 0, chennai_merged.columns[[1] + list(range(5, chennai_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
150  Western Suburbs of Chennai           Pizza Place           Bus Station   
151  Western Suburbs of Chennai           Pizza Place           Bus Station   
152  Western Suburbs of Chennai           Pizza Place           Bus Station   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
150                  Café  Fast Food Restaurant         Historic Site   
151                  Café  Fast Food Restaurant         Historic Site   
152                  Café  Fast Food Restaurant         Historic Site   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
150         Movie Theater     Indian Restaurant    Light Rail Station   
151         Movie Theater     Indian Restaurant    Light Rail Station   
152         Movie Theater     Indian Restaurant    Light Rail Station   

    9th Most Common Venue         10th Most Common Venue  \
150            Restaurant  Vegetarian / Vegan Restaurant   
151            Restaurant  Vegetarian / Vegan Restaurant   
152            Restaurant  Vegetarian / Vegan Restaurant   

    11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
150          Train Station         Breakfast Spot            Gas Station   
151          Train Station         Breakfast Spot            Gas Station   
152          Train Station         Breakfast Spot            Gas Station   

    14th Most Common Venue 15th Most Common Venue  
150         Sandwich Place                   Bank  
151         Sandwich Place                   Bank  
152         Sandwich Place                   Bank

# Results <a id ='results'></a>

The above analysis of Chennai concludes that the Western Suburbs of Chennai are considered to be most popular neighborhood that contains multiple places like cafe, fastfood restaurants, gas station, movie theatre etc. Although the south east Chennai region was closely competing, the western suburb of Chennai, especially Alwarthirunagar and Alapakkam turned out triumph with the number of amenities. From this, we can understand that these areas in particular are quite famous among the Chennai population and would be a good neighborhood to settle down.

# Discussion <a id ='discussion'></a>

In our analysis of the neighborhoods to settle down in Chennai, we have come across various features of the Chennai City. There are 187 unique categories of recreation and amenities available in various parts of chennai which indicates that Chennai has ample opportunities to grow into one of the leading cities of the world. We were also able to identify that the western region of chennai is more famous when compared to the other neighborhoods with the most common vendues including pizza place, bus station and cafe. 

From this we can identify that Chennai is open to trying out various cuisines from around the world and if anyone who settles in Chennai is fond of pizzas or is expert at making them, this is the place for them to be. 

There are a couple of other top venues that are eateries which includes that Chennai people indulge themselves in various food related activities.

Movie Theatre is a common recreation in and around Chennai and it is one of the most common venues in Western Chennai. People aspiring to settle down in the area do not have to worry about recreational activities. 

# Conclusion <a id ='conclusion'></a>

The goal of this project was to identify localities or neighborhood that is most popular in Chennai city for people who are aspiring to make Chennai their home. By analyzing the amenities and the quality of the amenities available in each area, we were able to arrive at a conclusion that the western suburbs are one of the top places for new people to settle down. 

There were a couple of other neighborhoods that were also closely popular and the people are spoilt for choices to pick their area.Final decision is left to the people who can make use of this analysis to find the right locality to live based on the most famous amenities and recreation facilities.
